src/main/java/org/opensearch/knn/index/vectorvalues/KNNVectorValues.java

```diff
@@ -23,11 +23,24 @@ public abstract class KNNVectorValues<T> {
 
     protected final KNNVectorValuesIterator vectorValuesIterator;
     protected int dimension;
+    protected long bytesPerVector;
 
     protected KNNVectorValues(final KNNVectorValuesIterator vectorValuesIterator) {
         this.vectorValuesIterator = vectorValuesIterator;
     }
 
+    /**
+     * Iterates the values once only if docIds is not at start position
+     * Also populates dimension and bytesPerVector in the process
+     * @throws IOException
+     */
+    public void init() throws IOException {
+        if (docId() == -1) {
+            nextDoc();
```

Comment:
This is a side effect of the decision that we can only get dimensions once we get vectors. We cannot get vectors without calling `nextDoc`. Similar to the fact that the user needs to be aware of the fact that you need to iterate once to get dimensions, the user needs to be aware when to iterate once. Moving this to client leads to repeated code which adds to maintenance

We can rethink the name here but this seems to be the right place rather than a random utility

Reply:
> We can rethink the name here but this seems to be the right place rather than a random utility

I am not saying to put it in the utility, put it in the NativeIndexWriter. Apart from indexing code, none needs to know what is the dimension. BTW, even on indexing you can use FieldInfo to get the dimension of the vectors.

On checking further on the code this line: https://github.com/opensearch-project/k-NN/pull/1950/files#diff-75a39a358693703da64ecd93181897b1e20ad25f951f0f3813148c21257a999cR104

for binary vector datatype will give size of array * 8 dimension value. See my comment on the same line there you don't need dimension. Hence just putting this on the fact that we cannot get dimension else we read the vector doesn't seem a correct reason for having this function.